### PR TITLE
[Feature][Connector-V2] StarRocks source connector

### PR DESCRIPTION
**Pull request by 531651225**
## Purpose of this pull request
refer to https://github.com/apache/incubator-seatunnel/issues/3018
add starrocks thrift source connector  
- the of Implementation principle starRocks thrift source connector refer to
https://docs.starrocks.io/zh-cn/latest/unloading/Flink_connector

# Benchmarking（starRocks thrift source connector）

## Hardware Environment & Software Environment

- Starrocks Version: 1.19.5
- Starrocks Cluster Info: 4 be nodes and  3 fe nodes

## Test Data
| table     | Record count | Column count |
|-----------|--------------|--------------|
| lineorder | 59986052     | 11           |

## Test Procedure
Configure seatunnel task to batch synchronize 60 million data from starrocks to starrocks, using 5 parallelism.

### Generate Data & Create Table Schema & Import Data
Refer to https://docs.starrocks.io/zh-cn/latest/benchmarking/SSB_Benchmarking

图片

图片

### define Job Config File
图片

## Test Results

图片

图片

图片

Configure seatunnel to batch synchronize 60 million data from starrocks to starrocks, using 5 parallelism.
- Total Job Runtime: 2m39s
- Starrocks Thrift Source connector operator Runtime: 1m5s

## Check list

* [x] Code changed are covered with tests, or it does not need tests for reason:
* [x] If any new Jar binary package adding in your PR, please add License Notice according
  [New License Guide](https://github.com/apache/incubator-seatunnel/blob/dev/docs/en/contribution/new-license.md)
* [x] If necessary, please update the documentation to describe the new feature. https://github.com/apache/incubator-seatunnel/tree/dev/docs
* [x] If you are contributing the connector code, please check that the following files are updated:
  1. Update change log that in connector document. For more details you can refer to [connector-v2](https://github.com/apache/incubator-seatunnel/tree/dev/docs/en/connector-v2)
  2. Update [plugin-mapping.properties](https://github.com/apache/incubator-seatunnel/blob/dev/plugin-mapping.properties) and add new connector information in it
  3. Update the pom file of [seatunnel-dist](https://github.com/apache/incubator-seatunnel/blob/dev/seatunnel-dist/pom.xml)

### CODE REVIEW

File: seatunnel-connectors-v2/connector-starrocks/src/main/java/org/apache/seatunnel/connectors/seatunnel/starrocks/client/source/StarRocksRowBatchReader.java [hunk 1]
```diff
@@ -0,0 +1,288 @@
...
+            }
+        } catch (Exception e) {
+            close();
+            throw new StarRocksConnectorException(StarRocksConnectorErrorCode.READER_ARROW_DATA_FAILED, e);
+        }
+    }
+
+    public SeaTunnelRow next() {
+        if (!hasNext()) {
+            String errMsg = "Get row offset:" + offsetInRowBatch + " larger than row size: " + readRowCount;
```
Comment (TaoZex) [inline seatunnel-connectors-v2/connector-starrocks/src/main/java/org/apache/seatunnel/connectors/seatunnel/starrocks/client/source/StarRocksRowBatchReader.java]: How about use String.format?

File: seatunnel-e2e/seatunnel-connector-v2-e2e/connector-starrocks-e2e/src/test/resources/starrocks-to-starrocks.conf [hunk 2]
```diff
@@ -45,4 +62,7 @@ sink {
         sink.properties.strip_outer_array = true
         max_retries = 3
     }
+    Console {
+
+    }
```
Comment (hailin0) [inline seatunnel-e2e/seatunnel-connector-v2-e2e/connector-starrocks-e2e/src/test/resources/starrocks-to-starrocks.conf]: Add `Assert` Sink check rows & fields

File: seatunnel-connectors-v2/connector-starrocks/src/main/java/org/apache/seatunnel/connectors/seatunnel/starrocks/client/source/StarRocksRowBatchReader.java (hunk 1 above)
Comment (531651225) [reply seatunnel-connectors-v2/connector-starrocks/src/main/java/org/apache/seatunnel/connectors/seatunnel/starrocks/client/source/StarRocksRowBatchReader.java]: > How about use String.format?

thanks，already fix it

File: seatunnel-e2e/seatunnel-connector-v2-e2e/connector-starrocks-e2e/src/test/resources/starrocks-to-starrocks.conf (hunk 2 above)
Comment (531651225) [reply seatunnel-e2e/seatunnel-connector-v2-e2e/connector-starrocks-e2e/src/test/resources/starrocks-to-starrocks.conf]: > Add `Assert` Sink check rows & fields

thanks，StarRocksIT contains Sink check rows & fields

Comment (531651225) [reply seatunnel-e2e/seatunnel-connector-v2-e2e/connector-starrocks-e2e/src/test/resources/starrocks-to-starrocks.conf]: added Benchmarking（starRocks thrift source connector) in comment

File: docs/en/connector-v2/source/StarRocks.md [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+
+| name                    | type   | required | default value     |
+|-------------------------|--------|----------|-------------------|
+| node_urls               | list   | yes      | -                 |
+| username                | string | yes      | -                 |
+| password                | string | yes      | -                 |
+| database                | string | yes      | -                 |
+| table                   | string | yes      | -                 |
+| scan_filter             | string | no       | -                 |
+| fields                  | config | yes      | -                 |
```
Comment (CalvinKirs) [inline docs/en/connector-v2/source/StarRocks.md]: For the newly added Connector, we use the scheam field to define this information. FYI https://github.com/apache/incubator-seatunnel/issues/3823

Comment (CalvinKirs) [reply docs/en/connector-v2/source/StarRocks.md]: here

Comment (531651225) [reply docs/en/connector-v2/source/StarRocks.md]: > For the newly added Connector, we use the scheam field to define this information. FYI #3823
thanks，already fix it.PTAL

File: docs/en/connector-v2/source/StarRocks.md [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+Read external data source data through StarRocks.
+The internal implementation of StarRocks source connector is obtains the query plan from the frontend (FE), 
+delivers the query plan as a parameter to BE nodes, and then obtains data results from BE nodes..
+## Key features
+
+- [x] [batch](../../concept/connector-v2-features.md)
+- [ ] [stream](../../concept/connector-v2-features.md)
+- [x] [exactly-once](../../concept/connector-v2-features.md)
+- [x] [schema projection](../../concept/connector-v2-features.md)
+
```
Comment (EricJoy2048) [inline docs/en/connector-v2/source/StarRocks.md:19]: image

You lost some key features.

File: docs/en/connector-v2/source/StarRocks.md [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+| name                    | type   | required | default value     |
+|-------------------------|--------|----------|-------------------|
+| node_urls               | list   | yes      | -                 |
+| username                | string | yes      | -                 |
+| password                | string | yes      | -                 |
+| database                | string | yes      | -                 |
+| table                   | string | yes      | -                 |
+| scan_filter             | string | no       | -                 |
+| fields                  | config | yes      | -                 |
```
Comment (EricJoy2048) [inline docs/en/connector-v2/source/StarRocks.md]: Please use `schema` options

File: seatunnel-connectors-v2/connector-starrocks/pom.xml [hunk 6]
```diff
@@ -54,5 +61,57 @@
...
+    <build>
+        <plugins>
+            <plugin>
+                <groupId>org.apache.maven.plugins</groupId>
+                <artifactId>maven-shade-plugin</artifactId>
+                <configuration>
+                    <relocations>
+                        <relocation>
+                            <pattern>org.apache.arrow</pattern>
+                            <shadedPattern>com.starrocks.shaded.org.apache.arrow</shadedPattern>
```
Comment (EricJoy2048) [inline seatunnel-connectors-v2/connector-starrocks/pom.xml]: Please use this format
image

File: seatunnel-connectors-v2/connector-starrocks/src/main/java/org/apache/seatunnel/connectors/seatunnel/starrocks/client/source/StarRocksBeReadClient.java [hunk 7]
```diff
@@ -0,0 +1,180 @@
...
+                if (!TStatusCode.OK.equals(result.getStatus().getStatus_code())) {
+                    throw new StarRocksConnectorException(StarRocksConnectorErrorCode.SCAN_BE_DATA_FAILED, "Failed to get next from be -> ip:[" + ip + "] "
+                            + result.getStatus().getStatus_code() + " msg:" + result.getStatus().getError_msgs());
+                }
+                eos.set(result.isEos());
+                if (!eos.get()) {
+                    rowBatch = new StarRocksRowBatchReader(result, seaTunnelRowType).readArrow();
+                }
+            } catch (TException e) {
+                throw new RuntimeException(e.getMessage());
```
Comment (EricJoy2048) [inline seatunnel-connectors-v2/connector-starrocks/src/main/java/org/apache/seatunnel/connectors/seatunnel/starrocks/client/source/StarRocksBeReadClient.java]: Please use the connector custom Exception.

**Review by EricJoy2048 — COMMENTED** [review]
Please add connector to `/seatunnel-dist/pom.xml` and `/config/plugin_config` file.

**Comment by 531651225** [comment]
> Please add connector to `/seatunnel-dist/pom.xml` and `/config/plugin_config` file.

thanks, It should already exist.
image
image

File: docs/en/connector-v2/source/StarRocks.md (hunk 4 above)
Comment (531651225) [reply docs/en/connector-v2/source/StarRocks.md:19]: > image
> 
> You lost some key features.

already fix it.PTAL

File: docs/en/connector-v2/source/StarRocks.md (hunk 5 above)
Comment (531651225) [reply docs/en/connector-v2/source/StarRocks.md]: already fix it.

File: seatunnel-connectors-v2/connector-starrocks/pom.xml (hunk 6 above)
Comment (531651225) [reply seatunnel-connectors-v2/connector-starrocks/pom.xml]: already fix it.

File: seatunnel-connectors-v2/connector-starrocks/src/main/java/org/apache/seatunnel/connectors/seatunnel/starrocks/client/source/StarRocksBeReadClient.java (hunk 7 above)
Comment (531651225) [reply seatunnel-connectors-v2/connector-starrocks/src/main/java/org/apache/seatunnel/connectors/seatunnel/starrocks/client/source/StarRocksBeReadClient.java]: already fix it.

**Comment by EricJoy2048** [comment]
LGTM, Please resolve conflicts.

**Comment by 531651225** [comment]
> LGTM, Please resolve conflicts.

thinks, fix it.PTAL

**Review by EricJoy2048 — DISMISSED** [review]

File: seatunnel-e2e/seatunnel-connector-v2-e2e/connector-starrocks-e2e/src/test/resources/starrocks-thrift-to-starrocks-streamload.conf [hunk 8]
```diff
@@ -21,12 +21,31 @@ env {
 }
 
 source {
-  Jdbc {
```
Comment (hailin0) [inline seatunnel-e2e/seatunnel-connector-v2-e2e/connector-starrocks-e2e/src/test/resources/starrocks-thrift-to-starrocks-streamload.conf:24]: remove jdbc from this maven pom

File: docs/en/connector-v2/source/StarRocks.md [hunk 9]
```diff
@@ -0,0 +1,151 @@
...
+The internal implementation of StarRocks source connector is obtains the query plan from the frontend (FE),
+delivers the query plan as a parameter to BE nodes, and then obtains data results from BE nodes.
+
+## Key features
+
+- [x] [batch](../../concept/connector-v2-features.md)
+- [ ] [stream](../../concept/connector-v2-features.md)
+- [x] [exactly-once](../../concept/connector-v2-features.md)
+- [x] [schema projection](../../concept/connector-v2-features.md)
+- [x] [parallelism](../../concept/connector-v2-features.md)
```
Comment (hailin0) [inline docs/en/connector-v2/source/StarRocks.md:17]: how to use

File: docs/en/connector-v2/source/StarRocks.md [hunk 10]
```diff
@@ -0,0 +1,151 @@
...
+
+Read external data source data through StarRocks.
+The internal implementation of StarRocks source connector is obtains the query plan from the frontend (FE),
+delivers the query plan as a parameter to BE nodes, and then obtains data results from BE nodes.
+
+## Key features
+
+- [x] [batch](../../concept/connector-v2-features.md)
+- [ ] [stream](../../concept/connector-v2-features.md)
+- [x] [exactly-once](../../concept/connector-v2-features.md)
```
Comment (hailin0) [inline docs/en/connector-v2/source/StarRocks.md]: how to use

**Review by hailin0 — APPROVED** [review]

**Review by CalvinKirs — APPROVED** [review]

**Pull request merged**